netbsd/NetBSDProcessList.c

```diff
@@ -0,0 +1,315 @@
+
+   for (int i = 0; i < count; i++) {
+      const struct kinfo_proc2* kproc = &kprocs[i];
+
+      bool preExisting = false;
+      Process* proc = ProcessList_getProcess(&this->super, kproc->p_pid, &preExisting, NetBSDProcess_new);
+
+      proc->show = ! ((hideKernelThreads && Process_isKernelThread(proc)) || (hideUserlandThreads && Process_isUserlandThread(proc)));
+
+      if (!preExisting) {
+         proc->ppid = kproc->p_ppid;
+         proc->tpgid = kproc->p_tpgid;
+         proc->tgid = kproc->p_pid;
+         proc->session = kproc->p_sid;
+         proc->tty_nr = kproc->p_tdev;
+         proc->pgrp = kproc->p__pgid;
+         proc->st_uid = kproc->p_uid;
+         proc->starttime_ctime = kproc->p_ustart_sec;
+         Process_fillStarttimeBuffer(proc);
+         proc->user = UsersTable_getRef(this->super.usersTable, proc->st_uid);
+         ProcessList_add(&this->super, proc);
+         proc->comm = NetBSDProcessList_readProcessName(this->kd, kproc, &proc->basenameOffset);
+      } else {
+         if (settings->updateProcessNames) {
+            free(proc->comm);
+            proc->comm = NetBSDProcessList_readProcessName(this->kd, kproc, &proc->basenameOffset);
+         }
+      }
+
+      proc->m_virt = kproc->p_vm_vsize;
+      proc->m_resident = kproc->p_vm_rssize;
+      proc->percent_mem = (proc->m_resident * pageSizeKB) / (double)(this->super.totalMem) * 100.0;
+      proc->percent_cpu = CLAMP(getpcpu(kproc), 0.0, this->super.cpuCount * 100.0);
+      proc->nlwp = kproc->p_nlwps;
+      proc->nice = kproc->p_nice - 20;
+      proc->time = 100 * (kproc->p_rtime_sec + ((kproc->p_rtime_usec + 500000) / 1000000));
+      proc->priority = kproc->p_priority - PZERO;
+
+      struct kinfo_lwp* klwps = kvm_getlwps(this->kd, kproc->p_pid, kproc->p_paddr, sizeof(struct kinfo_lwp), &nlwps);
+
+      switch (kproc->p_realstat) {
+      case SIDL:     proc->state = 'I'; break;
+      case SACTIVE:
+	// We only consider the first LWP with a one of the below states.
+        for (int j = 0; j < nlwps; j++) {
+          if (klwps) {
+            switch (klwps[j].l_stat) {
+            case LSONPROC: proc->state = 'P'; break;
+            case LSRUN:    proc->state = 'R'; break;
+            case LSSLEEP:  proc->state = 'S'; break;
+            case LSSTOP:   proc->state = 'T'; break;
+            default:       proc->state = '?';
+            }
+            if (proc->state != '?')
+            break;
+	  }
+	}
+        break;
+      case SSTOP:    proc->state = 'T'; break;
+      case SZOMB:    proc->state = 'Z'; break;
+      case SDEAD:    proc->state = 'D'; break;
+      default:       proc->state = '?';
+      }
+
+      this->super.totalTasks++;
+      // SRUN ('R') means runnable, not running
+      if (proc->state == 'P') {
+         this->super.runningTasks++;
+      }
+      proc->updated = true;
+   }
+}
+
+static unsigned long long saturatingSub(unsigned long long a, unsigned long long b) {
+   return a > b ? a - b : 0;
+}
+
```

Comment:
If this is used in multiple places, we may consider to extract this into `XUtils.[ch]` for generic re-use. This function may be of interest for other platforms too. Please extract as separate commit and update other platforms that may profit from this.

Reply:
This is also used by OpenBSD, I guess I should coordinate with the person who is responsible for the OpenBSD parts to do the refactor. Should I do it as a part of another ticket?

Reply:
If you do it as a separate commit that only refactors that aspect I'd even accept it as part of this series. Refactoring changes just shouldn't be mixed in with other stuff in the same commit. If the changes become too large we can still split it off into a separate PR.

Reply:
Understood, let me look into this then. Thank you for the advice.

Reply:
Should be refactored in d43cd5c